crates/matrix-sdk-ui/src/timeline/controller/decryption_retry_task.rs

```diff
@@ -137,42 +156,98 @@ async fn decryption_task<D: Decryptor>(
     debug!("Decryption task stopping.");
 }
 
-/// Return a list of the items within the timeline that we should retry
-/// decrypting because their session updated. Items are identified by their
-/// index in the supplied `state`'s list of items.
-async fn item_indices_to_retry(
-    state: Arc<RwLock<TimelineState>>,
+/// Decide which events should be retried, either for re-decryption, or, if they
+/// are already decrypted, for re-checking their encryption info.
+///
+/// Returns a tuple `(retry_decryption_indices, retry_info_indices)` where
+/// `retry_decryption_indices` is a list of the indices of UTDs to try
+/// decrypting, and retry_info_indices is a list of the indices of
+/// already-decrypted events whose encryption info we can re-fetch.
+fn event_indices_to_retry_decryption<'a>(
+    items: impl Iterator<Item = &'a Arc<TimelineItem>>,
```

Comment:
Can we pass something akin to `&[&Arc<TimelineItem>]` instead? Passing an iterator is bleh, and will negatively impact compile times.